### ISSUE TITLE
Program external link is unescaped in northstar UI

### ISSUE THREAD

**Issue by auzlee** [civiform/civiform#8762]
The external program link is unescaped here:
https://github.com/civiform/civiform/blob/a5ac309d85cd87ee813893837c6fc16d90ae36ce/server/app/views/applicant/IneligibleTemplate.html#L32

The only validation we do is to ensure it starts with `http://`. I'm not sure if that's enough to prevent xss:
https://github.com/civiform/civiform/blob/a5ac309d85cd87ee813893837c6fc16d90ae36ce/server/app/services/program/ProgramService.java#L686-L688

This link is only configurable by admins, so the risk is minimal, but it'd be good to not start this pattern.

**Comment by ssandbekkhaug** [comment]
Good news, our CSP prevents this kind of attack.

When I change the href to `javascript:alert('XSS example')`, the browser console logs the following error:
```
Refused to run the JavaScript URL because it violates the following Content Security Policy directive: "script-src 'nonce-RnvknUcynByeUHepSGRlZg==' 'strict-dynamic' https: 'unsafe-inline'". Note that 'unsafe-inline' is ignored if either a hash or nonce value is present in the source list.
```

**Comment by ssandbekkhaug** [comment]
I’m looking for a string to repro a XSS vulnerability that would be caught by escaping but not CSP. Does such a string exist?

**Comment by avaleske** [comment]
It looks like it might not, which is a good reason to keep our CSP!

That said, it'd still be good to fix this #DefenseInDepth

**Comment by ssandbekkhaug** [comment]
Totally agree we should fix this. The reason I want a string to repro is so that I have something to test against. Otherwise I don't know if my future change works.

**Comment by avaleske** [comment]
Can you turn off the CSP on your local to test it?

**Comment by ssandbekkhaug** [comment]
Good idea. That works for testing. I can repro an example of an XSS attack with `javascript:alert('XSS')` as a link href.

Next question: How do we want to escape strings?

- Import a library such as `org.apache.commons.text.StringEscapeUtils`
    - Pros: Well-vetted by security experts, zero maintenance cost
    - Cons: Adds yet another dependency to CiviForm
- Write our own function
    - Pros: Might be "good enough", avoids adding a new dependency
    - Cons: Might*not* be secure enough

**Comment by ssandbekkhaug** [comment]
Gwen recommended using TextFormatter class.

We're putting untrusted URLs in a href.

According to [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#xss-prevention-rules-summary), we should use the following mechanisms to prevent XSS attacks:

- [ ] Canonicalize input
- [ ] URL Validation
- [ ] Safe URL verification
- [ ] Allow-list http and HTTPS URLs only (Avoid the JavaScript Protocol to Open a new Window)
- [ ] Attribute encoder

Given that these URLs are supplied by admins, I'm lowering the priority of this issue to P1

**Comment by ssandbekkhaug** [comment]
After multiple attempts at URL encoding and sanitization, and discussing with other engineers, we have determined this is a non-trivial task.

- URLs cannot be encoded by Java's URLEncoder. You have to encode each piece individually. This makes it very difficult to encode a user-supplied URL.
- For example, encoding `https://en.wikipedia.org/w/index.php?search=aobeusaeou&title=Special%3ASearch&ns0=1` by naively using URLEncoder breaks the URL
- PR: https://github.com/civiform/civiform/pull/8829
- https://blog.hubspot.com/marketing/parts-url
- https://www.talisman.org/%7Eerlkonig/misc/lunatech%5Ewhat-every-webdev-must-know-about-url-encoding/#HandlingURLscorrectlyinJava

This issue is will be put in the backlog because:
- We have CSP. CSP provides robust protection against XSS attacks.
- The URLs are supplied by admins

**Comment by dkatzz** [comment]
@osajulian-dev are you working on this? If not, could you unassign. If you did it, please link the related PR
